tests/test_layabout.py

```diff
@@ -1 +1,327 @@
+        pass
+
+    @layabout.handle('hello', kwargs=kwargs)
+    def handle_hello4(slack, event, spam):
+        pass
+
+    layabout.handle(type='hello')(fn=handle_hello1)
+    layabout.handle(type='hello', kwargs=kwargs)(fn=handle_hello3)
+
+
+def test_layabout_raises_type_error_with_invalid_handlers():
+    """
+    Test that layabout raises a TypeError if an event handler is supplied that
+    doesn't meet the minimum criteria to be called correctly.
+    """
+    layabout = Layabout(name='test')
+
+    def invalid_handler1(slack):
+        pass
+
+    with pytest.raises(TypeError) as exc1:
+        layabout.handle(type='hello')(fn=invalid_handler1)
+
+    with pytest.raises(TypeError) as exc2:
+        @layabout.handle('hello')
+        def invalid_handler2(slack):
+            pass
+
+    expected = ("Layabout event handlers take at least 2 positional "
+                "arguments, a slack client and an event")
+    assert str(exc1.value) == expected and str(exc2.value) == expected
+
+
+def test_layabout_handlers_can_still_be_used_normally():
+    """
+    Test that layabout can decorate event handlers that can still be used as
+    though they were undecorated. Most importantly, that they can still return
+    and that their docstrings are intact.
+    """
+    layabout = Layabout(name='test')
+    cheese_shop = dict(shop='🧀')
+
+    @layabout.handle('hello')
+    def handle_hello(slack, event):
+        """ This docstring must be preserved. """
+        return cheese_shop
+
+    assert handle_hello(None, None) == cheese_shop
+    assert handle_hello.__doc__ == ' This docstring must be preserved. '
+
+
+def test_layabout_can_connect_to_slack_with_token(monkeypatch):
+    """
+    Test that layabout can connect to the Slack API when given a valid Slack
+    API token.
+    """
+    layabout = Layabout(name='test')
+
+    rtm_connect = MagicMock(return_value=True)
+    slack = MagicMock(rtm_connect=rtm_connect)
+    SlackClient = MagicMock(return_value=slack)
+    monkeypatch.setattr('layabout.SlackClient', SlackClient)
+
+    layabout.run(token=TOKEN, until=lambda e: False)
+
+    # Verify we instantiated a SlackClient with the given token and used it to
+    # connect to the Slack API.
+    SlackClient.assert_called_with(token=TOKEN)
+    rtm_connect.assert_called_with()
+
+
+def test_layabout_can_connect_to_slack_with_env_var(monkeypatch):
+    """
+    Test that layabout can discover and use a Slack API token from an
+    environment variable when not given one directly.
+    """
+    env_var = '_TEST_SLACK_API_TOKEN'
+    environ = {env_var: TOKEN}
+    layabout = Layabout(name='test', env_var=env_var)
+
+    rtm_connect = MagicMock(return_value=True)
+    slack = MagicMock(rtm_connect=rtm_connect)
+    SlackClient = MagicMock(return_value=slack)
+    monkeypatch.setattr(os, 'environ', environ)
+    monkeypatch.setattr('layabout.SlackClient', SlackClient)
+
+    # Purposefully don't provide a token so we have to use an env var.
```

Comment:
This comment might connect with the following line more if the following line passed `token=None`, but that might break the scenario you are trying to test.

Reply:
Nah, that'd be just fine. It's more explicit and that's OK by me.